### PR TITLE
Feature/mat 3278

### PR DESCRIPTION
**Pull request by RohitKandimalla**
[MAT-3278](https://jira.cms.gov/browse/MAT-3278)

Added a recursive call to add all children included libraries to fhirBundle.
Updated test cases as needed.

### CODE REVIEW

File: mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java [hunk 1]
```diff
@@ -144,7 +142,7 @@ private FhirIncludeLibraryResult createScratchResponse() {
     private FhirIncludeLibraryResult createMatGlobalCommonFunctionsFhir4Response() {
         FhirIncludeLibraryResult result = buildResult(matGlobalCommonFunctionsFhir4Library);
 
-        result.setLibraryReferences(new ArrayList<>());
+        result.setLibraryReferences(new HashSet<>());
```
Comment (ghost) [inline mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java:145]: Is order important?
If so, a LinkedHashSet can be used instead

**Review by ghost — DISMISSED** [review]

File: mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java [hunk 2]
```diff
@@ -51,40 +56,55 @@ private void checkIsFhir(CqlTextParser cqlTextParser) {
...
             try {
                 Bundle bundle = fetchBundle(include);
-                libraryReferences.setSearchResult(true);
-                libraryReferences.setReferenceEndpoint(bundle.getEntry().get(0).getFullUrl());
-                libraryReferences.setLibrary((Library) bundle.getEntry().get(0).getResource());
+
+                Library library = (Library) bundle.getEntry().get(0).getResource();
+
+                fhirIncludeLibraryReferences.setSearchResult(true);
+                fhirIncludeLibraryReferences.setReferenceEndpoint(bundle.getEntry().get(0).getFullUrl());
```
Comment (jkotanchik-SB) [inline mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java:87]: Does `Library` have a `getFullUrl()` method? If so, could it be used in place of re-navigating the bundle object?

File: mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java [hunk 3]
```diff
@@ -51,40 +56,55 @@ private void checkIsFhir(CqlTextParser cqlTextParser) {
...
         boolean result = true;
 
         LibraryProperties libraryProperties = cqlTextParser.getLibrary();
 
-        res.setLibraryName(libraryProperties.getName());
-        res.setLibraryVersion(libraryProperties.getVersion());
+        fhirIncludeLibraryResult.setLibraryName(libraryProperties.getName());
+        fhirIncludeLibraryResult.setLibraryVersion(libraryProperties.getVersion());
+
+        getIncludedLibraryReferences(cqlTextParser, fhirIncludeLibraryResult, result);
```
Comment (jkotanchik-SB) [inline mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java]: The `boolean result` will not be modified. Instead, how about updating `outcome` directly on the `FhirIncludeLibraryResult` object?
```suggestion
        getIncludedLibraryReferences(cqlTextParser, fhirIncludeLibraryResult);
```

File: mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java [hunk 4]
```diff
@@ -51,40 +56,55 @@ private void checkIsFhir(CqlTextParser cqlTextParser) {
...
+                    Attachment cqlAttachment = findCqlAttachment(library, "text/cql");
+                    String data = new String(cqlAttachment.getData(), StandardCharsets.UTF_8);
+                    var cqlTextParserIncludedLibrary = new CqlTextParser(data);
+                    getIncludedLibraryReferences(cqlTextParserIncludedLibrary, fhirIncludeLibraryResult, result);
+                }
             } catch (Exception e) {
                 log.debug("Error when processing include: {}", include);
-                libraryReferences.setSearchResult(false);
+                fhirIncludeLibraryReferences.setSearchResult(false);
                 result = false;
```
Comment (jkotanchik-SB) [inline mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java]:
```suggestion
                fhirIncludeLibraryReferences.setOutcome(false);
```

File: mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java [hunk 5]
```diff
@@ -51,40 +56,55 @@ private void checkIsFhir(CqlTextParser cqlTextParser) {
         }
     }
 
-
     private FhirIncludeLibraryResult processParser(CqlTextParser cqlTextParser) {
-        FhirIncludeLibraryResult res = new FhirIncludeLibraryResult();
+        var fhirIncludeLibraryResult = new FhirIncludeLibraryResult();
 
         boolean result = true;
```
Comment (jkotanchik-SB) [inline mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java]:
```suggestion
        fhirIncludeLibraryResult.setOutcome(true);
```

File: mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java (hunk 1 above)
Comment (RohitKandimalla) [reply mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java:145]: Do we? I am not aware of any order that is required while bundling the libraries. Although as of right now, the measure library will be at the top and then all included libraries are added thereafter.

**Review by jkotanchik-SB — CHANGES_REQUESTED** [review]
Pretty sure the `boolean` primitive won't be updated in the new `getIncludedLibraryReferences` method. Suggest modifying the `outcome` property instead.

File: mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java (hunk 1 above)
Comment (jkotanchik-SB) [reply mat-fhir-services/src/test/java/gov/cms/mat/fhir/services/service/packaging/LibraryIncludeProcessorTest.java:145]: >  the measure library will be at the top and then all included libraries are added thereafter.

That's all that is expected. The ordering of the included libraries isn't critical.

File: mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java (hunk 2 above)
Comment (RohitKandimalla) [reply mat-fhir-services/src/main/java/gov/cms/mat/fhir/services/components/fhir/FhirIncludeLibraryProcessor.java:87]: The library doesn't have a method 'getFullUrl'

**Review by jkotanchik-SB — APPROVED** [review]

**Pull request merged**